### PR TITLE
support iot workbench to configure devkit in remote development

### PR DESCRIPTION
**Pull request by RongXiefdu**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -59,6 +60,8 @@
     "adm-zip": "^0.4.13",
     "copy-paste": "^1.3.0",
     "impor": "^0.1.1",
+    "lodash": "^4.17.15",
```
Comment (Sneezry) [inline package.json]: Is `lodash` required here?

File: src/extension.ts [hunk 2]
```diff
@@ -13,11 +15,50 @@ const clipboard = impor(
...
+  const serialportOpen = vscode.commands.registerCommand(
+    'iotcube.serialportOpen',
+    (comPort: string, option: PortOption) => {
+      return serialport.SerialPortCtrl.open(comPort, option);
+    }
+  );
+
+  const serialportChooseCOM = vscode.commands.registerCommand(
+    'iotcube.serialportChooseCOM',
+    async (board: Board | undefined) => {
```
Comment (Sneezry) [inline src/extension.ts]: Use `board?: Board` for optional parameter

File: src/models/Interfaces/PortOption.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export interface PortOption {
+    baudRate: number | undefined;
```
Comment (Sneezry) [inline src/models/Interfaces/PortOption.ts]: Use `baudRate?: number` for optional member

File: src/models/serialportctrl.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
+import * as vscode from 'vscode';
+import * as _ from 'lodash';
```
Comment (Sneezry) [inline src/models/serialportctrl.ts]: Avoid importing the entire `lodash`, it's very very huge. Use its member which you need instead

File: src/models/serialportctrl.ts [hunk 5]
```diff
@@ -0,0 +1,182 @@
+import * as vscode from 'vscode';
+import * as _ from 'lodash';
+import * as os from 'os';
+import {Board} from './Interfaces/Board';
+import {PortOption} from './Interfaces/PortOption';
+
+const serialport = require('../../vendor/node-usb-native').SerialPort;
```
Comment (Sneezry) [inline src/models/serialportctrl.ts]: Naming style. Use `SerialPort` as class name

File: src/models/serialportctrl.ts [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            reject(err);
+          } else {
+            resolve(ports);
+          }
+        });
+      }
+    );
+  }
+
+  static async chooseCOM(board: Board | undefined): Promise<string> {
```
Comment (Sneezry) [inline src/models/serialportctrl.ts]: Optional parameter

File: src/models/serialportctrl.ts [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        reject: (reason: Error) => void
+      ) => {
+        let comList;
+        try{
+          comList = await SerialPortCtrl.getComList();
+        } catch (err) {
+          return reject(err);
+        }
+
+        const az3166 = board;
```
Comment (Sneezry) [inline src/models/serialportctrl.ts]: Serial port is a general utility class, and should not contain any specific device information. Move it to high level code

File: src/models/serialportctrl.ts [hunk 8]
```diff
@@ -20,70 +18,36 @@ export class SerialPortCtrl {
...
-          } else {
-            return false;
+            ports.forEach( async (port) => {
+              const com = {
+                "comName": port.comName,
+                "productId": port.productId,
+                "vendorId": port.vendorId
+              };
+              try {
+                await portList.push(com);
```
Comment (Sneezry) [inline src/models/serialportctrl.ts]: Why do we need `await` here? `push` should not be `async` method

File: src/models/serialportctrl.ts [hunk 9]
```diff
@@ -20,70 +18,36 @@ export class SerialPortCtrl {
     return os.platform();
   }
 
-  static getComList(): Promise<SerialPortInfo[]> {
+  static async getComList() {
     return new Promise(
       (
-        resolve: (value: SerialPortInfo[]) => void,
+        resolve: (value: Object) => void,
```
Comment (Sneezry) [inline src/models/serialportctrl.ts]: Use specific type here instead of `Object`

**Comment by dilin-MS** [comment]
>   // context.subscriptions.push(serialportSend);

Can remove useless comments.

---
Refers to: src/extension.ts:321 in 6c727cb. [](commit_id = 6c727cbd9779ad5e9ae46f0a09b36eaa29d5914c, deletion_comment = False)

**Comment by dilin-MS** [comment]
>   // const serialportOpen = vscode.commands.registerCommand(

Remove useless comments here.

---
Refers to: src/extension.ts:197 in 6c727cb. [](commit_id = 6c727cbd9779ad5e9ae46f0a09b36eaa29d5914c, deletion_comment = False)

File: src/extension.ts [hunk 10]
```diff
@@ -13,11 +14,50 @@ const clipboard = impor(
   './models/clipboard'
 ) as typeof import('./models/clipboard');
 
+const serialport = impor(
+  './models/serialportctrl'
+) as typeof import('./models/serialportctrl');
+
 // const serialport = impor(
```
Comment (dilin-MS) [inline src/extension.ts]: Remove useless comments here.

File: src/extension.ts (hunk 2 above)
Comment (RongXiefdu) [reply src/extension.ts]: Fixed

File: src/models/Interfaces/PortOption.ts (hunk 3 above)
Comment (RongXiefdu) [reply src/models/Interfaces/PortOption.ts]: Fixed.

File: src/models/serialportctrl.ts (hunk 4 above)
Comment (RongXiefdu) [reply src/models/serialportctrl.ts]: Fixed.

File: src/models/serialportctrl.ts (hunk 5 above)
Comment (RongXiefdu) [reply src/models/serialportctrl.ts]: Fixed.

File: src/models/serialportctrl.ts (hunk 6 above)
Comment (RongXiefdu) [reply src/models/serialportctrl.ts]: Fixed.

File: src/models/serialportctrl.ts (hunk 7 above)
Comment (RongXiefdu) [reply src/models/serialportctrl.ts]: Moved to IoT Device Workbench.

File: src/models/serialportctrl.ts (hunk 8 above)
Comment (RongXiefdu) [reply src/models/serialportctrl.ts]: Fixed.

File: src/models/serialportctrl.ts (hunk 9 above)
Comment (RongXiefdu) [reply src/models/serialportctrl.ts]: Fixed.

File: src/extension.ts (hunk 10 above)
Comment (RongXiefdu) [reply src/extension.ts]: Fixed. Please see commit: bec5698ba7299f92c0c0b13ff7b0a37674fe774d

**Pull request merged**